src/tutorial/output.md

```diff
@@ -100,6 +100,41 @@ to make your (and your user's) life easier.
 
 </aside>
 
+## A note on printing performance
+
+Printing to the terminal is surprisingly slow!
+If you call things like `println!` in a loop,
+it can easily become a bottleneck in an otherwise fast program.
+To speed this up,
+there's two things you can do.
+
+First,
+it helps to acquire a lock on `stdout` (or `stderr`)
+and use `writeln!` to print to it directly.
+This prevents the system from locking an unlocking `stdout` over and over again.
+
+Second,
+you might want to reduce the number of writes
+you actually "flush" to the terminal.
+`println!` tells the system to write to the terminal _every_ time,
+because it is usual to print each new line.
+If you don't need that,
+you can wrap your `stdout` handle in a [`BufWriter`]
+which by default buffers up to 8kB.
+(You can still call `.flush()` on this `BufWriter`
+when you want to print immediately.)
+
+```rust
+use std::io::{self, Write};
+
+let stdout = io::stdout(); // get the global stdout entity
+let handle = stdout.lock(); // acquire a lock on it
+let mut handle = io::BufWriter::new(handle); // optional: wrap that handle in a buffer
+writeln!(handle, "foo: {}", 42); // add `?` if you care about errors here
+```
+
+[`BufWriter`]: https://doc.rust-lang.org/1.29.0/std/io/struct.BufWriter.html
+
```

Comment:
Yeah, to be honest, I don't know that I would even include an example that locks stdout at all. If you're going to wrap stdout in a buffered writer, then unless you explicitly want to prevent output interleaving from other threads, then I don't really see a reason to hand a locked stdout to BufWriter. I think that if we show example code, we should _at least_ show the thing that most folks will want: either standard `println!` or `BufWriter`. If you still want to show locking after that, then I think that's fine, but not to the exclusion of `io::BufWriter::new(io::stdout())`.

Anyway, that's what I would do if I were writing this. :)

Reply:
To make this more complication, there are other reasons for not wanting to use `println!`. For example, it will panic on _any_ I/O error. That's generally not desired in "real" CLI programs, and to make matters worse, you actually usually need to do case analysis on the I/O error. That is, a broken pipe error should cause the program to stop what it's doing and exit gracefully where as any other I/O should be treated as a normal error condition.